### PR TITLE
feat(MonitoredTripController): add support for otp2 to checkItinerary

### PR DESCRIPTION
**Pull request by daniel-heppner-ibigroup**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR adds OTP2 support to the checkItinerary endpoint. It does this with a new query parameter added to the URL to specify the OTP version. The default version is 1 to ensure backwards compatibility.

### CODE REVIEW

**Comment by daniel-heppner-ibigroup** [comment]
I am open to better ways to do this, I am not very familiar with the code base as a whole. We could also add otp version to the config since I don't believe we'll ever have a middleware instance being used by two different versions at the same time. 

Also, we will eventually need to upgrade this to use GraphQL and improve the accuracy of this endpoint. This is just a stopgap so there's no regression when we move to OTP2.

File: src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java [hunk 1]
```diff
@@ -3,10 +3,12 @@
 import io.github.manusant.ss.ApiEndpoint;
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.mongodb.client.model.Filters;
+import io.github.manusant.ss.descriptor.ParameterDescriptor;
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java]: Remove unused import.

File: src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java [hunk 2]
```diff
@@ -1,6 +1,11 @@
 package org.opentripplanner.middleware.otp;
 
+import com.fasterxml.jackson.core.JsonProcessingException;
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java]: Remove unused import.

File: src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java [hunk 3]
```diff
@@ -25,4 +30,23 @@ public String uri() {
         return uri;
     }
     public String toString() { return name; }
+
+    public static OtpVersion intToVersion(int versionAsInt) {
+        switch(versionAsInt) {
+            case 1:
+            default:
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java:44]: `default` clause should be at the end of the `switch` statement for readability. `case 1:` will obvs have to move as well.

File: src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java [hunk 4]
```diff
@@ -25,4 +30,23 @@ public String uri() {
...
+                return OTP1;
+            case 2:
+                return OTP2;
+        }
+    }
+
+    public static OtpVersion getOtpVersionFromRequest(Request request) {
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java:54]: Add method comment.

File: src/main/java/org/opentripplanner/middleware/utils/ConfigUtils.java [hunk 5]
```diff
@@ -171,8 +171,7 @@ public static boolean hasConfigProperty(String name) {
         return hasConfigProperty(envConfig, name);
     }
 
-    /**
-     * Returns true if the given config has the requested property.
+    /** * Returns true if the given config has the requested property.
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/utils/ConfigUtils.java]: Revert this change.

File: src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java [hunk 6]
```diff
@@ -25,4 +30,23 @@ public String uri() {
         return uri;
     }
     public String toString() { return name; }
+
+    public static OtpVersion intToVersion(int versionAsInt) {
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java]: Change the method to `private`. Consider changing this method name to `getOtpVersion` and the parameter to `int version`.

File: src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java [hunk 7]
```diff
@@ -25,4 +30,23 @@ public String uri() {
...
+            case 2:
+                return OTP2;
+        }
+    }
+
+    public static OtpVersion getOtpVersionFromRequest(Request request) {
+        try {
+            return intToVersion(Integer.parseInt(request.queryParamOrDefault("otpVersion", "1")));
+        } catch (NumberFormatException e) {
+            logMessageAndHalt(request, HttpStatus.BAD_REQUEST_400, "Error parsing otpVersion parameter", e);
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java]: Nit. Full stop.

**Review by br648 — CHANGES_REQUESTED** [review]
A few minor changes. Overall I think this is the best approach over using a config value.

**Review by br648 — APPROVED** [review]
Thanks for the changes. All good.

File: src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java [hunk 8]
```diff
@@ -25,4 +29,33 @@ public String uri() {
...
     public String toString() { return name; }
+
+    /**
+     * Converts integer to OTP Version object.
+     * @param version Version number as an integer
+     * @return OtpVersion object
+     */
+    private static OtpVersion getOtpVersion(int version) {
+        switch(version) {
+            case 2:
```
Comment (miles-grant-ibigroup) [inline src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java:41]: Don't we have an enum we can use here instead?

File: src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java [hunk 9]
```diff
@@ -46,6 +47,7 @@ protected void buildEndpoint(ApiEndpoint baseEndpoint) {
             .post(path("/checkitinerary")
                     .withDescription("Returns the itinerary existence check results for a monitored trip.")
                     .withRequestType(MonitoredTrip.class)
+                    .withQueryParam().withName("otpVersion").withDefaultValue("1").and()
```
Comment (miles-grant-ibigroup) [inline src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java]: Can we use the enum here too?

File: src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java [hunk 10]
```diff
@@ -25,4 +29,33 @@ public String uri() {
...
+    }
+
+    /**
+     * Gets an OTP version from the Spark request header
+     * @param request Spark request object
+     * @return OtpVersion specified in request
+     */
+    public static OtpVersion getOtpVersionFromRequest(Request request) {
+        try {
+            return getOtpVersion(Integer.parseInt(request.queryParamOrDefault("otpVersion", "1")));
```
Comment (miles-grant-ibigroup) [inline src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java:56]: Can we use the enum here as well?

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
I'm confused -- is this adding OTP2 support? I see that OTP2 as a version is being passed through correctly, but I'm not seeing what we're doing with that new field?

I'm also not entirely sold on using a url parameter to decide whether to use otp1 or otp2. Wouldn't this maybe make more sense as a config parameter? Do we expect one middleware to handle both at once?

File: src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java (hunk 8 above)
Comment (daniel-heppner-ibigroup) [reply src/main/java/org/opentripplanner/middleware/otp/OtpVersion.java:41]: I am not sure what you mean. The problem is that the value coming from the API is an int.

**Comment by daniel-heppner-ibigroup** [comment]
> I'm also not entirely sold on using a url parameter to decide whether to use otp1 or otp2. Wouldn't this maybe make more sense as a config parameter? Do we expect one middleware to handle both at once?

@miles-grant-ibigroup That's what I was referring to with this. I'll look into doing it in the config instead of with qp.

> I am open to better ways to do this, I am not very familiar with the code base as a whole. We could also add otp version to the config since I don't believe we'll ever have a middleware instance being used by two different versions at the same time.

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
Still unclear as to what the OTP2 parameter is actually doing -- what change does it make if you set it to OTP2? Also I wouldn't name the configuration key OTP_VERSION, since both are actively supported at the same time.

Finally, it's a nitpick but would it be possible to replace all `"1"` and `"2"` with `OtpVersion.OTP1`/`OtpVersion.OTP2`?

**Comment by daniel-heppner-ibigroup** [comment]
> since both are actively supported at the same time.

I made this change because you said:

> Wouldn't this maybe make more sense as a config parameter? Do we expect one middleware to handle both at once?

I think at this point I'm as confused as you are.

**Pull request closed without merging**